### ISSUE TITLE
Issues training the factored3d network

### ISSUE THREAD

**Issue by tonyromarock** [shubhtuls/factored3d#18]
Hi shubhtuls,

Sorry for the long issue, but my questions require some explaining beforehand.

I have been trying to reproduce the results you obtained for the factored object prediction, but I have been having problems with the loss function not decreasing during training. I wanted to ask if you had a similar experience when training the network. I will first describe my training setup and show some visualizations comparing my model's performance to your results. Afterwards I will specify my questions regarding the model.

### Training Setup and Results

Here are my loss functions for each of the training steps (9 epochs of box3d_base, 1 epoch dwr_base, 1 epoch of dwr_shape_ft) using the same hyperparameters as mentioned in [training.md](https://github.com/shubhtuls/factored3d/blob/master/docs/training.md#training-factored-3d-prediction-models). My splits ({train: 28280, val: 3769, test: 5643}) of the SUNCG dataset are slightly smaller than yours ({train: 31127, val: 4150, test: 6226}), but I would still expect to get similar results than you. My splits are not a subset of your provided splits and there is only a small overlap (around 10%-20%) between the houses in each of our validation and test sets. I use the pretrained autoencoder you provided as inital weights for the voxel decoder used in dwr_shape_ft. 
(**You can click on the images to increase their size in the browser**) 

![losses_box3d 2 -dwr_base-_dwr_shape_ft](https://user-images.githubusercontent.com/7185919/52638007-30022400-2ed1-11e9-9b8f-05ddd54daa9b.png)

I compared my trained model with your provided model from the [cachedir](https://www.dropbox.com/sh/dg82g87yvhkg9xk/AAD5Drgr0fzGhWPkYRVDf8Pza?dl=0&preview=cachedir.tar.gz) based on their precision-recall plots (using [pr_plots.py](https://github.com/shubhtuls/factored3d/blob/master/benchmark/suncg/pr_plots.py)). 
The first column is my model on my test set, the second column is your provided model on my test set, and the final column are the results you provided on your test set in the paper.

![pr-plots_mynetwork_providednetwork_reportedresults](https://user-images.githubusercontent.com/7185919/52638296-fb429c80-2ed1-11e9-89e7-1e2b1423cc46.png)

There is a dip in performance when comparing your model on my test set, but my model greatly underperforms when compared to your model.

Finally, here is also a qualitative comparison of the shape prediction performance of my model (**left**) to your model (**right**).

![shapepredictioncomparison](https://user-images.githubusercontent.com/7185919/52638620-a0f60b80-2ed2-11e9-9bd7-d0f05f63f376.png)

For many objects my model produces something like a mean shape (only occupied voxels in the center of the grid) and the output values are often close to 0, which can be seen by the realtively empty grid when setting the cutoff threshold for voxels at 0.5. In some cases my model produces similar results to yours (like the television prediction in the last row).

### Questions

1. Did you have the same experience of the loss function decreasing very little from one epoch to the next as shown in my loss curves above or did it decrease more steadily during you training process?

2. I noticed that training the network does not require a lot of GPU memory (around 3 GB on a GTX Titan X). I was wondering if you maybe increased the batch size (the [training.md](https://github.com/shubhtuls/factored3d/blob/master/docs/training.md#training-factored-3d-prediction-models) mentions a batch size of 8) for the final model you used for evaluation?

3. When using the same weighting for the loss components as in  [training.md](https://github.com/shubhtuls/factored3d/blob/master/docs/training.md#training-factored-3d-prediction-models), the shape loss only makes up a small component of the total loss (as seen in the loss curves above, where the shape loss is often greatly dominated by the rotation and translation loss). Was this also the case for you during training?

4. Is the model you used for the evaluation in the paper the same as the one you provided in your [cachedir](https://www.dropbox.com/sh/dg82g87yvhkg9xk/AAD5Drgr0fzGhWPkYRVDf8Pza?dl=0&preview=cachedir.tar.gz)?

5. Do you think that training the dwr_shape_ft for more than 1 epoch can increase the voxel decoder's shape prediction performance?

Sorry for the very long issue, but I hope you can give me some pointers on what might be the problems of my training procedure.

Best regards,
Peter Mortimer

**Comment by shubhtuls** [comment]
Hi Peter,
The training does seem a bit off.

1) Did you re-render the data at your own end? I've seen some cases where the node renderings people got were off due to random reasons. As a sanity check, could you compare the renderings you used to ones I have [here](https://www.dropbox.com/sh/fkurja2x04wkca5/AAD5WNdmxVkC88WEJ8jIpKtYa?dl=0).

2) Maybe some options in training were different. I have my training log and config files attached, and hopefully these might help debug.

[opts.log](https://github.com/shubhtuls/factored3d/files/2856948/opts.log)
[loss_log.txt](https://github.com/shubhtuls/factored3d/files/2856963/loss_log.txt)

**Comment by tonyromarock** [comment]
Hi Shubham,
thank you for the fast response.

Regarding the node renderings: I did re-render the data on my end. My node renderings do look different to yours. Here's a visualization comparing my node renderings to the yours.
(1st col.: input image, 2nd col.: my node renderings color mapped, 3rd col.: your node renderings color mapped)

![node_comparison](https://user-images.githubusercontent.com/7185919/52797110-78f2dd80-3075-11e9-9a31-2b783195d329.png)

Some of my node renderings are fairly similar to yours (see row 1), but many are very unreasonable, especially when comparing them to yours.

I used the **render_node_indices.py** script you mentioned in [preprocessing.md](https://github.com/shubhtuls/factored3d/blob/master/docs/preprocessing.md#rendering-layout-and-node-images) to prouce the node renderings. I will look into if I can produce more reasonable node renderings.

Have you made your node renderings or auxiliary data (e.g.: edgebox proposals, gt bboxes, amodal depth images) pubicly availabe somewhere?

This would be really helpful to be able to reproduce your results. Especially when the preprocessing seems to fail due to random reasons.

Best regards,
Peter Mortimer

**Comment by shubhtuls** [comment]
Hi Peter,
Did you compute the node renderings using mesa? Someone else reported that onscreen renderings did not work for them.

**Comment by tonyromarock** [comment]
Hi Shubham,
Back then when I did the preprocessing, I used the on-screen rendering. I will look into if I can setup the off-screen rendering (osmesa) and figure out if that is the cause for the rendering issues.

Is it possible for you to also provide your opts.log and loss_log.txt for the dwr_base and dwr_shape_ft training stage? 
The one you provided earlier from the box3d experiment was very insightful for me to see which components of my loss curve do not quite fit yet.

Best regards,
Peter Mortimer

**Comment by shubhtuls** [comment]
Hi Peter,
Link to logs [here](https://www.dropbox.com/sh/i7bm83lsoxqorty/AADB_yytm5C0Gcn3XJLcNfCAa?dl=0). Though since all training crucially depends on the node renderings, I'd say getting those to be correct first would be critical.

Thanks
- Shubham

**Comment by tonyromarock** [comment]
Hi Shubham,
Thank you for the additional logs. But yes, I am trying to get the SUNCG Toolbox to work with off screen rendering. I am using [this guide](https://gist.github.com/SiyuanQi/600d1ce536791b7a3bd2e59fdbe69e66) to install mesa with offscreen support and it seems to work, but I am not able to compile the SUNCG Toolbox while specifying my local install of osmesa. 

In the [preprocessing docs](https://github.com/shubhtuls/factored3d/blob/master/docs/preprocessing.md#compiling-suncg-toolbox) you mention changing some settings in [Makefile.apps](https://github.com/shurans/SUNCGtoolbox/blob/master/gaps/makefiles/Makefile.apps#L42), but I am not sure if I am doing this correctly. Can you let me know what you specified there exactly?

Thanks for the help,
Peter Mortimer

**Comment by shubhtuls** [comment]
Hi Peter,
To install the toolbox, I needed to add another path in that line. Although I think posting an issue in the [original repository](https://github.com/shurans/SUNCGtoolbox) for installation help might work better.

In the meantime, a temporary link to node renderings [here](https://www.dropbox.com/s/6aau6w5pkqo7w5f/renderings_node.tar.gz?dl=0). Hopefully you can reproduce things with this.
